rust/lance-encoding/src/encodings/physical/buffers.rs

```diff
@@ -48,7 +64,230 @@ impl BufferEncoder for BitmapBufferEncoder {
+            bitpack_meta: None,
+        };
         Ok(buffer)
     }
 }
+
+#[derive(Debug, Default)]
+pub struct BitpackingBufferEncoder {}
+
+impl BufferEncoder for BitpackingBufferEncoder {
+    fn encode(&self, arrays: &[ArrayRef]) -> Result<EncodedBuffer> {
+        let mut num_bits = 0;
+        for arr in arrays {
+            let arr_max = compute_num_bits(arr.clone())?;
+            num_bits = num_bits.max(arr_max);
+        }
+
+        // TODO handle case where there are no arrays or all the arrays are full of zeros
+        if num_bits == 0 {
+            panic!("TODO handle zero length bitpacking")
+        }
+
+        let mut packed_arrays = vec![];
+        for arr in arrays {
+            let packed = pack_array(arr.clone(), num_bits)?;
+            packed_arrays.push(packed.into());
+        }
+
+        let data_type = arrays[0].data_type();
+        let bits_per_value = 8 * data_type.byte_width() as u64;
+
+        Ok(EncodedBuffer {
+            bits_per_value: num_bits,
+            parts: packed_arrays,
+            bitpack_meta: Some(pb::BitpackMeta {
+                uncompressed_bits_per_value: bits_per_value,
+            }),
+        })
+    }
+}
+
+// TODO write some unit tests for this
+fn compute_num_bits(arr: ArrayRef) -> Result<u64> {
+    let max = match arr.data_type() {
+        DataType::UInt8 => num_bits_for_type::<UInt8Type>(arr.as_primitive()),
+        DataType::UInt16 => num_bits_for_type::<UInt16Type>(arr.as_primitive()),
+        DataType::UInt32 => num_bits_for_type::<UInt32Type>(arr.as_primitive()),
+        DataType::UInt64 => num_bits_for_type::<UInt64Type>(arr.as_primitive()),
+        _ => None,
+    };
+
+    max.ok_or(Error::InvalidInput {
+        source: format!(
+            "Invalid type supplied to compute bits for bitpacking: {}",
+            arr.data_type()
+        )
+        .into(),
+        location: location!(),
+    })
+}
+
+fn num_bits_for_type<T>(arr: &PrimitiveArray<T>) -> Option<u64>
+where
+    T: ArrowPrimitiveType,
+    T::Native: PrimInt + AsPrimitive<u64>,
+{
+    let max = arrow::compute::bit_or(arr);
+    max.map(|x| arr.data_type().byte_width() as u64 * 8 - x.leading_zeros() as u64)
+}
+
+fn pack_array(arr: ArrayRef, num_bits: u64) -> Result<Vec<u8>> {
+    match arr.data_type() {
+        DataType::UInt8 | DataType::UInt32 => Ok(pack_buffers(
+            arr.to_data(),
+            num_bits,
+            arr.data_type().byte_width(),
+        )),
+        _ => Err(Error::InvalidInput {
+            source: format!("Invalid data type for bitpacking: {}", arr.data_type()).into(),
+            location: location!(),
+        }),
+    }
+}
+
+fn pack_buffers(data: ArrayData, num_bits: u64, byte_len: usize) -> Vec<u8> {
+    let buffers = data.buffers();
+    let mut packed_buffers = vec![];
+    for buffer in buffers {
+        let packed_buffer = pack_bits(&buffer, num_bits, byte_len);
+        packed_buffers.push(packed_buffer);
+    }
+    packed_buffers.concat()
```

Comment:
We only want to pack the values buffer, I think this will also try and pack the validity buffer.

Reply:
I think we're actually OK here. This gets passed the result of `array.to_data()` here:
https://github.com/lancedb/lance/blob/d18b7dfc374e6c36aa35f39bb1f910b98c2e96b4/rust/lance-encoding/src/encodings/physical/bitpack.rs#L165-L168

And the validity buffer doesn't get included in that result. For example:
```rs
  let arr = UInt16Array::from(vec![Some(1), None, Some(2)]);
  let data = arr.to_data();
  let buffers = data.buffers();
  for buffer in buffers {
      println!("{:?}", buffer);
  }  
```
prints:
```
Buffer { data: Bytes { ptr: 0x124704e80, len: 6, data: [1, 0, 0, 0, 2, 0] }, ptr: 0x124704e80, length: 6 }
```